src/write.jl

```diff
@@ -1,140 +1,120 @@
         field = getfield(metadata, name)
         if T <: RecordingID && name === :startdate
-            bytes_written += write_bytes(io, "Startdate ")
-        end
-        bytes_written += write_bytes(io, field) + Base.write(io, 0x20)
-    end
-    return bytes_written
-end
-
-function write_bytes(io::IO, annotations::T) where {T<:Union{DataRecord, Vector{<:AbstractAnnotation}}}
-    bytes_written = 0
-    for annotation in annotations
-        bytes_written += write_bytes(io, annotation)
-    end
-    return bytes_written
-end
-
-function write_bytes(io::IO, annotation::AbstractAnnotation)
-    bytes_written = Base.write(io, sign(annotation)) +
-                    write_bytes(io, annotation.offset)
-    if has_duration(annotation)
-        bytes_written += Base.write(io, 0x15) +
-                         write_bytes(io, duration(annotation))
-    end
-    bytes_written += Base.write(io, footer(annotation)...)
-    if has_events(annotation)
-        for event in annotation.events
-            bytes_written += Base.write(io, event, 0x14)
+            bytes_written += _edf_write(io, "Startdate ")
         end
+        bytes_written += field isa Missing ? Base.write(io, 'X') : _edf_write(io, field)
+        bytes_written += Base.write(io, ' ')
     end
-    bytes_written += Base.write(io, 0x0)
     return bytes_written
 end
 
-sign(annotation::AbstractAnnotation) = annotation.offset >= 0 ? '+' : '-'
-
-has_duration(annotation::RecordAnnotation) = false
-has_duration(annotation::TimestampAnnotation) = annotation.duration !== nothing
-
-duration(annotation::RecordAnnotation) = nothing
-duration(annotation::TimestampAnnotation) = annotation.duration
-
-footer(annotation::RecordAnnotation) = (0x14, 0x14)
-footer(annotation::TimestampAnnotation) = 0x14
-
-has_events(annotation::RecordAnnotation) = annotation.events !== nothing
-has_events(annotation::TimestampAnnotation) = true
-
-function write_padded(io::IO, value, byte_limit::Integer; pad::UInt8=0x20)
-    bytes_written = write_bytes(io, value)
+function edf_write(io::IO, value, byte_limit::Integer; pad::UInt8=UInt8(' '))
+    bytes_written = _edf_write(io, value)
     bytes_written <= byte_limit || error("Written value $value contains more bytes than limit $byte_limit")
     while bytes_written < byte_limit
         bytes_written += Base.write(io, pad)
     end
     return bytes_written
 end
 
+#####
+##### `write_header`
+#####
+
+const BYTES_PER_FILE_HEADER = 256
+
+const BYTES_PER_SIGNAL_HEADER = 256
+
 function write_header(io::IO, file::File)
-    has_annotations = file.annotations !== nothing
-    signal_count = length(file.signals) + has_annotations
-    bytes_written = write_file_header(io, file.header, signal_count) +
-                    write_signal_headers(io, file, has_annotations)
-    reserved_bytes = 32 * signal_count
-    bytes_written += write_padded(io, 0x20, reserved_bytes)
+    bytes_written = BYTES_PER_FILE_HEADER + BYTES_PER_SIGNAL_HEADER * length(file.signals)
```

Comment:
Thought: it might be more intuitive to name this `expected_header_size_in_bytes` or something similar, then rename `actual_bytes_written` to `bytes_written`, since it's the only value actually tracking bytes written.

Reply:
good call!

Reply:
Upon doing this, it becomes apparent that we can just get rid of the `expected_header_size_in_bytes` and just inline the value into the `@assert` statement. Just pushed that change. Nice!